### PR TITLE
Common table components

### PR DESCRIPTION
**Pull request by IgorZiegler**
- Add table-cell component which represents as td tag
 - Add table-row component which contains all table-cells
 - Add table-rows component which renders all table-rows
 - Add table-header component which renders table headers
 - Add vp-table component which contains header and body component
 - Add tests for all components

### CODE REVIEW

File: app/components/vp-table/template.hbs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<table>
```
Comment (maxmalov) [inline app/components/vp-table/template.hbs]: should be a `tagName` in `vp-table`

File: app/fixtures/provider.json [hunk 2]
```diff
@@ -0,0 +1,44 @@
+{
```
Comment (maxmalov) [inline app/fixtures/provider.json]: I think it will be more meaningful to give a name to each column, and specify each row as a plain object with properties mapped over column names, like:

```js
columns: [{
  name: 'user',
  title: 'Пользовательэ
}, ...],

rows: [{
  user: 'Иванов Сергей' // this will be placed in the first column
  // ...
}, ...]
```

Comment (maxmalov) [inline app/fixtures/provider.json]: Besides, each plain table can be describes with its columns and rows, so please keep this in mind and update table components ;)

File: tests/integration/components/table-header/component-test.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  this.set('headers', [
+    'Пользователь',
+    'Тип',
+    'Значение',
+    'Дата'
+  ]);
+  this.render(hbs`{{table-header headers=headers}}`);
+
+  assert.equal(this.$(testSelector('th')).length, this.get('headers').length);
+});
```
Comment (maxmalov) [inline tests/integration/components/table-header/component-test.js]: how about title assertions?

File: tests/integration/components/table-row/component-test.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    cells: [
+      'Сидоров Иван',
+      'Комиссия с клиента',
+      '8',
+      '22/01/2015 16:02'
+    ]
+  });
+  this.render(hbs`{{table-row row=rowCells}}`);
+
+  assert.equal(this.$(testSelector('table-cell')).length, this.get('rowCells').cells.length);
```
Comment (maxmalov) [inline tests/integration/components/table-row/component-test.js]: same as [here](https://github.com/hwdtech/vp-dashboard-ember/pull/5/files#r75160596)

**Comment by maxmalov** [comment]
I guess we can decompose table a little bit in a different way. I propose the following schema:

```handlebars

{{!-- usage of table --}}
{{#table}}
  {{table-header columns=columns}}
  {{table-rows columns=columns rows=rows}}
{{/table}}
```

Thus adds customization to table inner component, for example if we don't wan't to show table header, we just skip it

File: app/components/vp-table/template.hbs (hunk 1 above)
Comment (IgorZiegler) [inline app/components/vp-table/template.hbs]: @maxmalov I use div wrapper for padding, because `border-collapse: collapse;` doesn't work with padding.

File: app/components/table-header/style.scss [hunk 5]
```diff
@@ -0,0 +1,5 @@
+th {
```
Comment (abukurov) [inline app/components/table-header/style.scss]: use class selector instead of tags

File: app/components/table-header/template.hbs [hunk 6]
```diff
@@ -0,0 +1,5 @@
+<tr>
+  {{#each headers as |header|}}
+    <th data-test-th>{{header.title}}</th>
```
Comment (abukurov) [inline app/components/table-header/template.hbs]: `data-test-table-header-row`

File: tests/integration/components/table-header/component-test.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  name: 'type',
+  title: 'Тип'
+}, {
+  name: 'value',
+  title: 'Значение'
+}, {
+  name: 'Date',
+  title: 'Дата'
+}];
+const renderHeaders = (context, headers = defaultHeaders) => {
```
Comment (abukurov) [inline tests/integration/components/table-header/component-test.js]: `renderTableHeader`

File: tests/integration/components/table-header/component-test.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  context.render(hbs`{{table-header headers=headers}}`);
+};
+
+test('should renders all table headers', function (assert) {
+  renderHeaders(this);
+  assert.equal(this.$(testSelector('th')).length, this.get('headers').length);
+});
+
+test('should renders table titles correctly', function (assert) {
+  renderHeaders(this);
```
Comment (abukurov) [inline tests/integration/components/table-header/component-test.js]: `assert.expect(defaultHeaders.length)`

File: tests/integration/components/table-header/component-test.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  this.$(testSelector('th')).each((index, header) => {
```
Comment (abukurov) [inline tests/integration/components/table-header/component-test.js]: `{}` is redudant

File: tests/integration/components/table-header/component-test.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    assert.ok(this.get('headers').find(item => this.$(header).text() === item.title));
```
Comment (abukurov) [inline tests/integration/components/table-header/component-test.js]: 'findBy(property, value)'

File: tests/integration/components/table-row/component-test.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+});
+
+test('should render all table cells', function (assert) {
+  renderTableRow(this);
+  assert.equal(this.$(testSelector('table-cell')).length, Object.keys(this.get('rowCells')).length);
+});
+
+test('should renders table cells correctly', function (assert) {
+  renderTableRow(this);
+  this.get('tableColumns').forEach((header) => {
```
Comment (abukurov) [inline tests/integration/components/table-row/component-test.js]: use `assrt.expect`

File: tests/integration/components/table-rows/component-test.js [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+}
+];
+const defaultTableRows = [{
+  user: 'Иванов Сергей',
+  type: 'Фиксированная комиссия с клиента',
+  value: '13',
+  Date: '22/01/2015 16:02'
+}];
+const renderTableRow = (context, headers = defaultHeaders, rows = defaultTableRows) => {
+  context.set('headers', headers);
```
Comment (abukurov) [inline tests/integration/components/table-rows/component-test.js]: `.setProperties`

File: app/components/table-header/template.hbs (hunk 6 above)
Comment (IgorZiegler) [inline app/components/table-header/template.hbs]: Did you mean `data-test-table-header-cell`?

File: app/components/table-row/template.hbs [hunk 13]
```diff
@@ -0,0 +1,3 @@
+{{#each headers as |header|}}
```
Comment (maxmalov) [inline app/components/table-row/template.hbs]: There is one header but multiple columns, so let's  rename it

File: tests/integration/components/table-row/component-test.js [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+test('should render all table cells', function (assert) {
+  renderTableRow(this);
+  assert.equal(this.$(testSelector('table-cell')).length, Object.keys(this.get('rowCells')).length);
+});
+
+test('should renders table cells correctly', function (assert) {
+  renderTableRow(this);
+  const tableColumns = this.get('tableColumns');
+  assert.expect(tableColumns.length);
```
Comment (maxmalov) [inline tests/integration/components/table-row/component-test.js]: What is this assertion checking? I mean you've set tableColumns in the test, and now check that it was set.

File: tests/integration/components/table-rows/component-test.js [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+  name: 'user',
+  title: 'Пользователь'
+}, {
+  name: 'type',
+  title: 'Тип'
+}, {
+  name: 'value',
+  title: 'Значение'
+}, {
+  name: 'Date',
```
Comment (maxmalov) [inline tests/integration/components/table-rows/component-test.js]: Why is this type capitalized?

File: tests/integration/components/table-rows/component-test.js [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+  title: 'Дата'
+}
+];
```
Comment (maxmalov) [inline tests/integration/components/table-rows/component-test.js]: Should be at the previous lube

Comment (maxmalov) [inline tests/integration/components/table-rows/component-test.js]: *line )))

File: tests/integration/components/vp-table/component-test.js [hunk 17]
```diff
@@ -0,0 +1,44 @@
+import { moduleForComponent, test } from 'ember-qunit';
+import hbs from 'htmlbars-inline-precompile';
+import testSelector from 'vp-dashboard-ember/tests/helpers/ember-test-selectors';
+const defaultHeaders = [{
```
Comment (maxmalov) [inline tests/integration/components/vp-table/component-test.js]: Looks like copy paste

File: tests/integration/components/table-rows/component-test.js [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+  context.render(hbs`{{table-rows headers=headers rows=rows}}`);
+};
+
+moduleForComponent('table-rows', 'Integration | Component | table rows', {
+  integration: true
+});
+
+test('should render all table rows', function (assert) {
+  renderTableRow(this);
+  assert.equal(this.$(testSelector('table-row')).length, this.get('rows').length);
```
Comment (maxmalov) [inline tests/integration/components/table-rows/component-test.js]: We also should check that table rows are corresponding to passed rows

File: tests/integration/components/table-row/component-test.js (hunk 14 above)
Comment (IgorZiegler) [inline tests/integration/components/table-row/component-test.js]: It's checks how many assertions are expected to run within a test.

File: tests/integration/components/table-rows/component-test.js (hunk 18 above)
Comment (IgorZiegler) [inline tests/integration/components/table-rows/component-test.js]: @maxmalov We are checked it in table-row test

File: app/provider/template.hbs [hunk 19]
```diff
@@ -1 +1,4 @@
-Provider route
\ No newline at end of file
+{{#content-title}}Поставщик «{{model.name}}»{{/content-title}}
+<div style="width: 51%; padding: 0.7em 2.4em; background-color: #f3f3f3;">
```
Comment (abukurov) [inline app/provider/template.hbs]: remove inline styles!!!

File: app/components/table-row/style.scss [hunk 20]
```diff
@@ -0,0 +1,7 @@
+& {
+  border-bottom: 1px solid #c1c5cB;
```
Comment (abukurov) [inline app/components/table-row/style.scss]: move color into variable

File: app/components/table-cell/style.scss [hunk 21]
```diff
@@ -0,0 +1,6 @@
+& {
+  font-size: 0.9em;
```
Comment (abukurov) [inline app/components/table-cell/style.scss]: `.9 * 16 = 14.4`

File: tests/integration/components/table-row/component-test.js [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+  renderTableRow(this);
+  assert.equal(this.$(testSelector('table-cell')).length, Object.keys(this.get('rowCells')).length);
+});
+
+test('should renders table cells correctly', function (assert) {
+  renderTableRow(this);
+  assert.expect(Object.keys(defaultRowCells).length);
+  this.get('columns').forEach((header) => {
+    assert.equal(this.$(testSelector('name', header.name)).text(),
+      this.get('rowCells')[header.name]);
```
Comment (abukurov) [inline tests/integration/components/table-row/component-test.js]: this.get(`rowCells.${header.name}`)

File: app/components/table-rows/component.js [hunk 23]
```diff
@@ -0,0 +1,7 @@
+import Ember from 'ember';
```
Comment (abukurov) [inline app/components/table-rows/component.js]: It's `table-body` component

File: tests/integration/components/vp-table/component-test.js [hunk 24]
```diff
@@ -0,0 +1,43 @@
+import { moduleForComponent, test } from 'ember-qunit';
+import hbs from 'htmlbars-inline-precompile';
+import testSelector from 'vp-dashboard-ember/tests/helpers/ember-test-selectors';
+const defaultColumns = [{
```
Comment (maxmalov) [inline tests/integration/components/vp-table/component-test.js]: 1 empty line after imports

File: tests/integration/components/vp-table/component-test.js [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+  name: 'type',
+  title: 'Тип'
+}, {
+  name: 'value',
+  title: 'Значение'
+}, {
+  name: 'date',
+  title: 'Дата'
+}];
+const defaultTableData = [{
```
Comment (maxmalov) [inline tests/integration/components/vp-table/component-test.js]: one line between multiline variables

File: tests/integration/components/table-rows/component-test.js (hunk 18 above)
Comment (maxmalov) [inline tests/integration/components/table-rows/component-test.js]: You're checking table cells there, not table rows

File: tests/integration/components/table-header/component-test.js [hunk 26]
```diff
@@ -0,0 +1,38 @@
+import { moduleForComponent, test } from 'ember-qunit';
+import hbs from 'htmlbars-inline-precompile';
+import testSelector from 'vp-dashboard-ember/tests/helpers/ember-test-selectors';
+
+moduleForComponent('table-header', 'Integration | Component | table header', {
+  integration: true
+});
+
+const defaultHeaders = [{
```
Comment (maxmalov) [inline tests/integration/components/table-header/component-test.js]: consts should be before `moduleForComponent` call

File: app/components/vp-table/template.hbs [hunk 27]
```diff
@@ -0,0 +1,2 @@
+{{table-header headers=columns}}
```
Comment (maxmalov) [inline app/components/vp-table/template.hbs]: https://github.com/hwdtech/vp-dashboard-ember/pull/5#issuecomment-240475754

File: tests/integration/components/table-rows/component-test.js (hunk 18 above)
Comment (IgorZiegler) [inline tests/integration/components/table-rows/component-test.js]: Done

File: app/components/table-rows/template.hbs [hunk 28]
```diff
@@ -0,0 +1,3 @@
+{{#each rows as |tableCells|}}
+  {{table-row cells=tableCells columns=columns}}
+{{/each}}
```
Comment (abukurov) [inline app/components/table-rows/template.hbs]: remove empty line

File: app/fixtures/provider.json [hunk 29]
```diff
@@ -0,0 +1,50 @@
...
+  "name": "Тестовый поставщик",
+  "isActive": true,
+  "history": {
+    "columns": [
+      {
+        "name": 'user',
```
Comment (abukurov) [inline app/fixtures/provider.json]: string should be double quoted

File: app/components/table-rows/template.hbs (hunk 28 above)
Comment (IgorZiegler) [inline app/components/table-rows/template.hbs]: Done

File: app/components/provider-status/style.scss [hunk 30]
```diff
@@ -3,9 +3,10 @@
   border-radius: 50%;
   width: 0.5em;
   height: 0.5em;
-  background-color: $provider-status-inactive-color;
+  background-color: #e56372;
```
Comment (maxmalov) [inline app/components/provider-status/style.scss]: should be a variable

File: tests/integration/components/table-header/component-test.js [hunk 31]
```diff
@@ -0,0 +1,39 @@
+import { moduleForComponent, test } from 'ember-qunit';
+import hbs from 'htmlbars-inline-precompile';
+import testSelector from 'vp-dashboard-ember/tests/helpers/ember-test-selectors';
+
+const defaultHeaders = [{
```
Comment (maxmalov) [inline tests/integration/components/table-header/component-test.js]: move this copypasted columns to a test fixture

**Comment by maxmalov** [comment]
Except a few remarks LGTM

File: app/app.js [hunk 32]
```diff
@@ -7,7 +7,7 @@ Ember.MODEL_FACTORY_INJECTIONS = true;
 
 const App = Ember.Application.extend({
   'ember-composable-helpers': {
-    only: ['join']
+    only: ['join', 'find-by']
```
Comment (maxmalov) [inline app/app.js]: Cannot see where `find-by` is used

File: app/components/provider-status/style.scss [hunk 33]
```diff
@@ -4,8 +4,8 @@
   width: 0.5em;
   height: 0.5em;
   background-color: $provider-status-inactive-color;
-  display: inline-block;
-  margin-bottom: 0.2em;
+  margin-bottom: 0.1em;
```
Comment (maxmalov) [inline app/components/provider-status/style.scss]: How is provider status related to table component? Please be consistent and leave unrelated code untouched

File: tests/fixtures/table-olumns.json [hunk 34]
```diff
@@ -0,0 +1,18 @@
+[
```
Comment (maxmalov) [inline tests/fixtures/table-olumns.json]: should be `table-columns`, not `table-olumns`

File: tests/integration/components/table-rows/component-test.js [hunk 35]
```diff
@@ -0,0 +1,39 @@
...
+});
+
+test('should render all table rows with correct content', function (assert) {
+  renderTableRow(this);
+  assert.expect(defaultColumns.length * defaultTableRows.length);
+
+  this.get('columns').forEach((header) => {
+    this.$(`${testSelector('table-row')} ${testSelector('column-name', header.name)}`)
```
Comment (abukurov) [inline tests/integration/components/table-rows/component-test.js]: `${testSelector('table-row')}:nth-child(${index + 1}) ${testSelector('column-name', header.name)}`

File: app/provider/services/template.hbs [hunk 36]
```diff
@@ -1 +1,4 @@
-<h1>{{model.name}}</h1>
\ No newline at end of file
+{{#vp-table}}
```
Comment (abukurov) [inline app/provider/services/template.hbs]: move into `kitchen-sink` template

File: app/fixtures/history.json [hunk 37]
```diff
@@ -0,0 +1,46 @@
+{
```
Comment (abukurov) [inline app/fixtures/history.json]: it is not histories fixtures

File: app/kitchen-sink/controller.js [hunk 38]
```diff
@@ -1,4 +1,40 @@
...
+    title: 'Значение'
+  }, {
+    name: 'Date',
+    title: 'Дата'
+  }],
+  rows: [{
+    user: 'Иванов Сергей',
+    type: 'Фиксированная комиссия с клиента',
+    value: '13',
+    Date: '22/01/2015 16:02'
```
Comment (maxmalov) [inline app/kitchen-sink/controller.js]: why this property is uppercased?
